### ISSUE TITLE
📣 Plan for Ruby 3.0 support (Released!)

### ISSUE THREAD

**Issue by ainame** [fastlane/fastlane#17931]
👋 Hello, fastlane community!

Ruby 3.0 support was released at v2.183.0. Please give it a try and give us feedback on this issue.
https://github.com/fastlane/fastlane/releases/tag/2.183.0

The plan is now hidden under "Details".
<details>

**Please react with :+1:  emoji on this issue if you are waiting for Ruby 3.0 support!**

This is @ainame who recently became a core contributor 🚀  I'm pleased to announce that I will be working on getting Ruby 3.0 support in fastlane. So this ticket is to track the progress and let fastlane users be aware of the work being addressed.

## TL; DR

* Use Ruby 2.4-2.7 until 3.0 migration is done
* Don't send us a bug report about installation failure with Ruby 3.0 as it will cause duplication a lot
* Help us if we need to test any for the migration in future :pray:

## Ruby 3.0 and currently supported versions

[Ruby 3.0 was released on 25th December 2020](https://www.ruby-lang.org/en/news/2020/12/25/ruby-3-0-0-released/). It's a major version update but its source code compatibility with 2.x should be good. There are performance improvements and interesting new features implemented.

However, fastlane hasn't officially supported Ruby 3.0 yet. Please read the below section for details. The currently supported versions are Ruby 2.4-2.7. I highly recommend that you upgrade Ruby version to 2.6 or 2.7. Ruby 2.4 is now in EOL period, which means there will be no security fix for it. Ruby 2.5 will be the same soon.

https://www.ruby-lang.org/en/downloads/branches/ 

If you try to install fastlane against Ruby 3.0, it actually succeeds with the version "2.54.1" BUT it is too old and `deliver` won't work for the latest App Store Connect APIs. Please don't use it.

### Ruby 2.7's warnings

Before going to Ruby 3.0, we should make sure the last version in 2.x series "2.7" is supported correctly. It's very important because 2.7 warns us if there are chances of breaking with some deprecation messages to prepare Ruby 3.0 migration. https://github.com/fastlane/fastlane/issues/15856

It's been recognised, in fact, fastlane can run on Ruby 2.7, which is the latest version in 2.x series. However, despite the fact, fastlane project hasn't guaranteed that all the testing specs pass with Ruby 2.7 on CI yet. This will be resolved on https://github.com/fastlane/fastlane/pull/17861

### Blocker to install fastlane to Ruby 3.0

Right now, we have the dependency `google-api-client` that we have a strict limit in the flexibility of version resolving.

>  spec.add_dependency('google-api-client', '>= 0.37.0', '< 0.39.0') # Google API Client to access Play Publishing API

However, within those versions, none of them allows us to install it to Ruby 3.0 env. We need to adopt 0.44.2 or later https://github.com/googleapis/google-api-ruby-client/pull/900.

### Replace unmaintained gem

There is also `slack-notifier` gem which *was* popular to send messages to Slack but however, it doesn't seem to be actively maintained. So we might need to replace that gem with other or one made by ourselves.

### To run unit tests

`rspec/spec-mocks` https://github.com/rspec/rspec-mocks/pull/1394 also still has some issues for Ruby 3.0. We need to wait for it to be fixed.

### Drop Ruby 2.4 support 

We'd like to drop 2.4 support when we officially release "Ruby 3.0" support. It's mainly because rubygems we are using also did the same. So the more rubygems do that the harder we keep supporting old versions; e.g. #18339. Essentially the Ruby core team has ended security support for 2.4 last year. So it's good to update your Ruby version to keep your environment secure.

googleapis/google-api-ruby-client#2841
googleapis/signet#178

## Todo

This is a rough plan to work on the migration.

1. ~Enable CI for Ruby 2.7~ https://github.com/fastlane/fastlane/pull/17861~
2. ~Prepare ~Danger~ a CI check to prevent master/main branch to cause regressions of 2.7 warnings before enabling Ruby 3.0 on CircleCI https://github.com/fastlane/fastlane/pull/18021~
3. Fix deprecation warnings with Ruby 2.7 if there are any to improve the source code compatibility with 3.0 
   * ~Fix warnings in the codebase https://github.com/fastlane/fastlane/pull/18021~
   * ~Replace `slack-notifier` with something else or implement a basic Slack integration by ourselves~ #18512  #18537
   * ~Wait for `rspec-mocks` to release a version that supports Ruby 3.0 or fix existing specs that crashes due to Ruby 3.0 syntax~  Apparently, we fixed this issue before I know it
4. Work on other migrations
   * ~Update rubocop version to the latest~ https://github.com/fastlane/fastlane/pull/18564
   * ~Update dependencies that have deprecation warnings; highline, etc...~ https://github.com/fastlane/fastlane/pull/18599
   * ~Bump `google-api-client` version to unblock Ruby 3.0 installation~ https://github.com/fastlane/fastlane/pull/18656
   * ~Fix failing specs due to rspec's syntax unless rspec team releases a new version that supports Ruby 3.0~
   * ~Enable CI for Ruby 3.0~ https://github.com/fastlane/fastlane/pull/18656
5. ~Release the version that supports Ruby 3.0 officially~

For 3-5, depending on how big impact changes will have, we might cut a work branch to stack up changes and release RC branch. If it will not be too big, changes will be merged into `master` gradually.

</details>

**Comment by ainame** [comment]
```
% rbenv 2.7.2
% bundle install
% RUBYOPT="-W:deprecated" bundle exec rspec --pattern */**/*_spec.rb 2> stderr.txt
% cat stderr.txt | grep warning: > warnings.txt
```

With those commands, I got some warnings that need to be fixed for Ruby 3.0 (including other normal warnings). 

https://gist.github.com/ainame/d7f1c4d2fb07f139d3a0f064d4cbda75

I'll be fixing them and sending PRs once #17861 gets merged. Maybe I should create a Danger lint rule to check if there are such warnings appearing on rspec output.

**Comment by ainame** [comment]
I'm accumulating fixes onto this branch.  
https://github.com/ainame/fastlane/commits/fix-ruby2.7-warnings

**Comment by mikehardy** [comment]
I recognize this is off-topic but it doesn't appear there are too many subscribers - just wanted to say thanks! open source maintenance is usually a thankless job but shouldn't be, and fastlane in particular is so useful to me, thanks for helping. Cheers @ainame

**Comment by ainame** [comment]
#17861 and #17892 got merged! We'll be able to test fastlane with Ruby 2.7 finally!!

**Comment by ainame** [comment]
This PR will be the one to 🔪 deprecated warnings as much as I found in fastlane itself and I could fix. 
https://github.com/fastlane/fastlane/pull/18021

However, I left this method due to the difficulty to fix. I guess this may involve a refactoring on call sites.

https://github.com/fastlane/fastlane/blob/633860a5aae0fa9c0b936797560efb24181b1bf7/fastlane_core/lib/fastlane_core/project.rb#L80

**Comment by ainame** [comment]
I realised that without keeping checking warnings, there may be regressions of my fixes for warnings. So I will work on Danger checks for it. Added this to the TODO list.

> Prepare Danger to prevent master branch to cause regressions of 2.7 warnings before enabling Ruby 3.0 on CircleCI

**Comment by ainame** [comment]
I've also started working on document pages to clarify the currently supported Ruby versions.
https://github.com/fastlane/docs/issues/1018

**Comment by RamblinWreck77** [comment]
Just here to cheer everyone on! Looking forward to a 3.0-compatible release. :)

**Comment by ainame** [comment]
This PR is now ready to review! 🚀 
https://github.com/fastlane/fastlane/pull/18021

It took me a while to implement a CI check that prevents regressions without using Danger. It turned out we won't be able to enable Danger for forked repositories as it can leak GitHub's access token.

**Comment by ainame** [comment]
https://github.com/fastlane/fastlane/pull/18021 has got merged. From now on no new code that is not ready for Ruby 3.0 in terms of "Separation of positional and keyword arguments" can be merged (as far as it's covered by unit tests).

And also @joshdholtz the lead maintainer and I discussed dropping Ruby 2.4 support. We'd like to drop 2.4 support when we officially release "Ruby 3.0" support. It's mainly because rubygems we are using also did the same. So the more rubygems do that the harder we keep supporting old versions. Essentially the Ruby core team has ended security support for 2.4 last year. So it's good to update your Ruby version to keep your environment secure.

https://github.com/googleapis/google-api-ruby-client/pull/2841
https://github.com/googleapis/signet/pull/178

**Comment by ainame** [comment]
While working on this improvement https://github.com/fastlane/fastlane/pull/18278 in order to find undiscovered issues, I could find another issue in the codebase for Ruby 3.0🙂
https://github.com/fastlane/fastlane/pull/18395

**Comment by ainame** [comment]
Replacement for `slack-notifier` is ready. I'll submit the PR shortly. 
https://github.com/ainame/fastlane/compare/30555fc...replace-slack-notifier

**Comment by ainame** [comment]
Bumping `rubocop` version to the latest is on the way...
https://github.com/fastlane/fastlane/compare/master...ainame:bump-rubocop-version

**Comment by ainame** [comment]
highline 1.7.x is not compatible with Ruby 3.0. I'll be migrating it to 2.x. 
https://github.com/JEG2/highline/pull/248

**Comment by ainame** [comment]
In order to update `highline` gem, we first need to bump `commander` by stop using the forked one https://github.com/fastlane/commande.
I will submit a couple of PRs for it one by one. 

https://github.com/fastlane/fastlane/pull/18599

**Comment by mikehardy** [comment]
You are on the [ultimate yak shave](https://www.techopedia.com/definition/15511/yak-shaving) @ainame :-), thanks so much for your work here

**Comment by ainame** [comment]
@mikehardy It's my pleasure. Yeah, it is yak shaving but that's the only way to do it right here😂 
I will keep things up-to-date or resolve tech-debt stuff for the next major change from time to time as long as I maintain fastlane😅

**Comment by ainame** [comment]
I will work on `google-api-client` migration. It seems like it's not only about version bumping but also we need to adopt the new way of distribution of the gem split by its service. I'll first investigate what Google services are used in fastlane.

https://github.com/googleapis/google-api-ruby-client/blob/master/google-api-client/OVERVIEW.md

**Comment by ainame** [comment]
It's getting really close to the goal😄 https://github.com/fastlane/fastlane/pull/18656

**Comment by ainame** [comment]
This PR is ready to review and test! I appreciate it if anyone can volunteer to test the below branch as it updates `google-api-client` gem. It shouldn't have any breaking changes but want to ensure that 🙇 

```
gem 'fastlane', git: 'https://github.com/fastlane/fastlane.git', branch: 'ai/migrate-google-api-client'
```
https://github.com/fastlane/fastlane/pull/18656

**Comment by mikehardy** [comment]
I wanted to help try this but I don't have a Gemfile for my fastlane installation, and when I attempt to create one on macOS 11.3.1 with homebrew gcc 10.2 installed (current stable?) it fails at the `bundle install` step

```
Installing json 2.5.1 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /Users/mike/.rvm/gems/ruby-2.7.2/gems/json-2.5.1/ext/json/ext/generator
/Users/mike/.rvm/rubies/ruby-2.7.2/bin/ruby -I /Users/mike/.rvm/rubies/ruby-2.7.2/lib/ruby/2.7.0 -r ./siteconf20210508-16778-n2z7k9.rb extconf.rb
creating Makefile

current directory: /Users/mike/.rvm/gems/ruby-2.7.2/gems/json-2.5.1/ext/json/ext/generator
make "DESTDIR=" clean

current directory: /Users/mike/.rvm/gems/ruby-2.7.2/gems/json-2.5.1/ext/json/ext/generator
make "DESTDIR="
compiling generator.c
gcc: error: unrecognized command-line option '-Wdivision-by-zero'; did you mean '-Wdiv-by-zero'?
gcc: error: unrecognized command-line option '-Wshorten-64-to-32'
gcc: error: unrecognized command-line option '-Wextra-tokens'; did you mean '-Wextra-semi'?
make: *** [generator.o] Error 1

make failed, exit code 2

Gem files will remain installed in /Users/mike/.rvm/gems/ruby-2.7.2/gems/json-2.5.1 for inspection.
Results logged to /Users/mike/.rvm/gems/ruby-2.7.2/extensions/x86_64-darwin-20/2.7.0/json-2.5.1/gem_make.out

An error occurred while installing json (2.5.1), and Bundler cannot continue.
Make sure that `gem install json -v '2.5.1' --source 'https://rubygems.org/'` succeeds before bundling.

```

If there is some trivial solution to this I'll try the install locally - but I don't intend to take any of your time troubleshooting what is a likely semi-broken local GCC / compiler vs ruby environment incompatibility

**Comment by mikehardy** [comment]
On a different machine a brew update today gave me a new gcc/llvm/etc (which was hinted at as a resolution for my issue in the comment I hid, above) and I was able to successfully create a Gemfile with your git source above and run `bundle update`

I ran `bundle exec fastlane <etc>` with your git source and it appeared to execute as before to build an iOS app and push it to the TestFlight servers.

I suspect that with the google-apis being the thing you really want to test, someone that uses Fastlane to interact with the Play Store would be your ideal test scenario yes? Hopefully someone else can give that a test, but I can verify at least for a typical TestFlight / iOS build scenario, there is no regression

Thanks for your continued effort in this area, looking forward to ruby 3 support when it's ready 😃

**Comment by ainame** [comment]
@mikehardy Thank you for giving it a try. Your feedback using my branch with real-world `Fastfile` is very important to me. Glad to hear it worked😄

**Comment by ainame** [comment]
Hello, there👋 It's going to be released tommorrow!
Thank you for waiting for Ruby 3.0 patiently.
https://twitter.com/joshdholtz/status/1393019142119231494?s=21

**Comment by ainame** [comment]
It's released! 🎉 
https://github.com/fastlane/fastlane/releases/tag/2.183.0

Note that there is a known issue where `rdoc` v6.3.1 causes fastlane's installation failure. You can work around it by either using `bundler` + `Gemfile` or `--no-doc` flag.
https://github.com/fastlane/fastlane/issues/18642

**Comment by ainame** [comment]
This issue might be related to the last change
https://github.com/fastlane/fastlane/issues/17931

**Comment by ainame** [comment]
There will be a breaking change in Ruby 3.1.0 due to `psych` bumped to v4.0.0, which has a breaking change for fastlane.
https://github.com/ruby/ruby/blob/3208a5df2dfb429752a130a36274464e9924cf44/ext/psych/lib/psych/versions.rb#L5

`psych` is a default gem bundled in Ruby itself. You can update the version if you want. The latest release was a major version bump and included the breaking change that affects `http-cookie` gem that fastlane depends on internally. 

I'll submit a PR to `http-cookie` repo to resolve the actual cause of the issue.

**Comment by ainame** [comment]
This will also be broken with psych v4.0. We need to use keyword arguments for `YAML.safe_load` 
https://github.com/fastlane/fastlane/blob/27dfa591e5de0bffed85aa5e9f59d2d7234fabe3/spaceship/lib/spaceship/spaceauth_runner.rb#L42-L47

**Comment by ainame** [comment]
It's been a while since released. I'm going to close this ticket.
Going forward, I'm planning to add CI checks against Ruby head in order to find issues like the above ahead.
